40 feat workload logic

Maxtho8 opened the pull request:
# Overall explanation of our work :

We implemented routes for managing workloads:

 - /workload/ (**GET**)   : Get all workloads
 - /workload/ (**PUT**) : Create a new workload
 - /workload/<workload_id> (**DELETE**) : Delete a workload by id
 - /workload/<workload_id> (**GET**) : Get a workload by id
 - /workload/<workload_id> (**PATCH**) : Update a workload


## Architecture 

The code is divided into two parts :

- ### Service
    It implements all the logic to interact with etcd ( insert , retrieve ...)
- ### Controller
    It contains the routes and the functions they call

A service can load and use other services , in our case WorkloadService use EtcdService. 
The controller use WorkloadService as middleware functions

## Tests 

- **Test 1 :**

Insert a workload and retrieve it with his id

- **Test 2 :**

Insert a workload and delete it

- **Test 3 :**

Get all workloads and check HttpResponse

## Used libraries 
- **serde_json** : serialise and deserialise Workload to JSON
- **uuid** : generate an uuid for workload

sameo commented:
Still a draft I guess? You really should rebase that PR sooner than later...

sameo commented on controller/lib/src/external_api/workload/controller.rs:
```diff
@@ -1,46 +1,318 @@
-    HttpResponse::Ok().body("put_workload")
-}
+impl WorkloadController {
+    pub fn get_services(&self) -> Scope {
```
`services`, not `get_services`.

sameo replied:
`services`, not `get_services`.

sameo commented on controller/lib/src/external_api/workload/service.rs:
```diff
@@ -1,17 +1,177 @@
+    /// Returns:
+    ///
+    /// A string
+    pub fn get_id(&mut self, name: &str, namespace: &str) -> String {
```
`id` not `get_id`

sameo commented on controller/lib/src/external_api/workload/controller.rs:
```diff
@@ -1,46 +1,318 @@
+                    _ => http_response.body("Internal Server Error"),
+                }
+            })
+            .unwrap();
```
No `unwrap`, please. Return a `Result` and propagate the error.

Maxtho8 replied:
In controller i can't propagate the error because it  needs to return HTTP response.
I use the unwrap because all error are handled line 43. If I replace unwrap by "?", the trait bound `actix_web::HttpResponse: ` is not satisfied ...

sameo commented on controller/lib/src/external_api/workload/controller.rs:
```diff
@@ -1,46 +1,318 @@
+            .get_workload(&workload_id, &namespace)
+            .await
+        {
+            Ok(workload) => match serde_json::to_string(&workload) {
+                Ok(json) => HttpResponse::build(StatusCode::OK).body(json),
+                Err(err) => HttpResponse::InternalServerError().body(format!(
+                    "Error while converting the workload to json string : {}",
+                    err
+                )),
+            },
```
This patterns repeats several times across this PR, you should create a `Workload` method that converts `self` to a `HttpResponse` (and take a `StatusCode` as an input parameter) and use that method instead of having several copies of that code snippet.

Maxtho8 replied:
I implemented the function to_http for Workload struct.
It serialize the workload in JSON, if failed return a http error, else it return the json in a http response.
What do you think about it ?

sameo commented on controller/lib/src/external_api/workload/controller.rs:
```diff
@@ -1,46 +1,318 @@
+                    err
+                )),
+            },
+            Err(e) => match e {
+                WorkloadError::WorkloadNotFound => {
+                    HttpResponse::build(StatusCode::NOT_FOUND).body("Workload not found")
+                }
+                WorkloadError::JsonToWorkload(msg) => {
+                    HttpResponse::build(StatusCode::INTERNAL_SERVER_ERROR).body(format!(
+                        "Error while converting JSON string to workload : {}",
+                        msg
+                    ))
+                }
+                _ => HttpResponse::build(StatusCode::INTERNAL_SERVER_ERROR)
+                    .body("Internal Server Error"),
+            },
+        }
+    }
```
Similar comment here: Please define and use a `WorkloadError`converter method. Since you don't need any specific parameter, you should use the `From`trait directly: https://doc.rust-lang.org/std/convert/trait.From.html

Maxtho8 replied:
Yes indeed, I implementedthe function `to_http` for WorkloadError enum. 
It returns the corresponding http error, What do you think about it ?

sameo commented on controller/lib/src/external_api/workload/controller.rs:
```diff
@@ -216,103 +216,111 @@ impl WorkloadController {
-    };
-    use serde_json;
-    use std::str::from_utf8;
+// #[cfg(test)]
+// mod tests {
+//     use super::WorkloadController;
+//     use actix_web::{
+//         http::{header::ContentType, StatusCode},
+//         test, web, App,
+//     };
```
This commit is not adding any tests, it's deleting them.
Why? When are those tests going to be re-enabled? Is there an issue tracking that? All of this should be part of your commit message at least.

Maxtho8 replied:
There were some changes that broke the tests. I spent a lot of time trying to patch them but I didn't succeed. As we have little time left I decided to delete the commit.

sameo replied:
It's ok. But your commit message should reflect that.

Maxtho8 replied:
So I put the tests back into a commit , saying that they don't work anymore? Or I put it in an existing commit ?

sameo reviewed (changes requested):
Thanks for the PR @Maxtho8.
Besides the code comments I added, please fix the following:

1. Commit messages: your first commit adds 600 lines of code with zero commit messqage and an elusive commit subject. Please expalin what you're doing, why you're doing it and how.
2. Smaller commits: Again, your first commit is pretty big. Split it at least in 2, the first one adds the service code and the second ones the controller implementation.

sameo commented on controller/lib/src/external_api/workload/controller.rs:
```diff
@@ -1,46 +1,318 @@
+    /// # Arguments:
+    ///
+    /// * `params`: web::Path<(String, String)> - The first Path parameter is the namespace and the second the workload id.
+    pub async fn get_workload(
```
`workload`, not `get_workload`

Maxtho8 committed:
doc  : add documentation

Signed-off-by: Maxime <maxime.pizzolitto.82@gmail.com>

Maxtho8 committed:
feat : add workload service


Workload service implements all the logic to manage workloads
(A service can load and use other services)
Workload service implements EtcdService which is the interface to interact with etcd.

Things to know : 

- Workload service require etcd address when initialized

- Each route handler create a new service , to avoid code repetition we use new_or_http_error function which automatically sends an http error if the service creation fails because of etcd. 

- Etcd save in key => value format.
The key is the namespace concatenated with the workload name

Signed-off-by: Maxime <maxime.pizzolitto.82@gmail.com>

sameo commented on controller/lib/src/external_api/workload/controller.rs:
```diff
@@ -1,46 +1,168 @@
+        match workload_service
+            .get_workload(&workload_id, &namespace)
+            .await
+        {
+            Ok(workload) => workload.to_http(),
+            Err(e) => e.to_http(),
+        }
```
This is `map_or_else(e e.to_http(), w w.to_http())?`

Maxtho8 replied:
this looks better however HttpResponse from actix does not implement the Try feature to use `?`

sameo replied:
You actually don't need the `?`. The `map_or_else`call should return `HttpResponse`, which implements `Responder`afaiu.

Maxtho8 replied:
yes my bad

Maxtho8 commented on controller/lib/src/external_api/workload/model.rs:
```diff
@@ -1,11 +1,48 @@
+}
+
+#[derive(Deserialize, Serialize)]
+#[derive(Deserialize, Serialize, Clone, Debug)]
```
I can't find the line you indicate...

sameo commented on controller/lib/src/external_api/workload/model.rs:
```diff
@@ -1,11 +1,87 @@
+}
+
+#[derive(Deserialize, Serialize)]
+
```
Unneeded new line.

Maxtho8 commented on controller/lib/src/external_api/workload/controller.rs:
```diff
@@ -1,46 +1,168 @@
+
+        let mut workload_service = WorkloadService::new_or_http_error(&data.etcd_address)
+            .await
+            .unwrap();
```
I'm going to replace this, but it won't make the code too repetitive?

sameo replied:
Repetitive is better than panic :-)

Maxtho8 replied:
Thanks , Wouldn't it be even better to do it this way  ?

```rust
let mut workload_service = match WorkloadService::new(&data.etcd_address) {
     Ok(workload) => workload,
     Err(e) => return  e.to_http(),```

sameo replied:
> Yes you are right , Wouldn't it be even better to do it this way ?
> 
> ```rust
> let mut workload_service = match WorkloadService::new(&data.etcd_address) {
>      Ok(workload) => workload,
>      Err(e) => return  e.to_http(),
> ```

Yes; that's better indeed.

> I would like to know why the unwrap could panic since in to_http the match handles all error cases

By definition `unwrap` can panic. If the function called can never return a `Error`, then it should not return a `Result`. If it does return a `Result` and you `unwrap` it, you're explicitly taking the risk of making your implementation panic.

Maxtho8 replied:
Ok thanks, I changed the code , it's better ?

Maxtho8 committed:
feat: add workload controller


Add routes and routes handler for workload.  
Controller use workload service.

- /workload/<namespace> (GET) : Get all workloads
- /workload/<namespace>  (PUT) : Create a new workload
- /workload/<namespace>/<workload_id> (DELETE) : Delete a workload by id
- /workload/<namespace>/<workload_id> (GET) : Get a workload by id
- /workload/<namespace>/<workload_id> (PATCH) : Update a workload

Things to know

- function `services` is called by the server to get all the routes it needs to handle. It associates routes with middleware functions

- pagination is used on this route : /workload/<namespace> 
offset : start from the element <offset>
limit : retrieve a maximum of <limit> elements 

- When using pagination in get all route ,  query argument are optional but if they are used it is necessary to specify the two arguments (limit and offset). It is not possible to specify a single argument on both .


Signed-off-by: Maxime <maxime.pizzolitto.82@gmail.com>
